### PR TITLE
Install function deps before zip + ship

### PR DESCRIPTION
**Pull request by DavidWells**
Automatically install sub function folder deps

### CODE REVIEW

File: src/install-deps.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+const cp = require("child_process");
+const parallelLimit = require("run-parallel-limit");
+const series = require("run-series");
+
+module.exports = function installFunctionDeps(src) {
+  const fns = functionGlobber(src)
+  return new Promise((resolve, reject) => {
+    if (fns.length === 0) return resolve()
+
+    const jobs = fns.map(fnPath => {
```
Comment (sw-yx) [inline src/install-deps.js]: worth filtering something here in case some function folders do not have a package.json?

Comment (bcomnes) [reply src/install-deps.js]: The globber should only select folders with package.json if I understand correctly:

https://github.com/netlify/zip-it-and-ship-it/pull/34/files#diff-f1580f02b18be88d05da772c352b9bbaR27

File: src/install-deps.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      return resolve(data)
+    })
+  })
+}
+
+function functionGlobber(baseDir) {
+  const globStr = path.join(baseDir, "*/package.json")
+  const functions = glob.sync(globStr)
+  return functions.map(fnFolder => {
+    return fnFolder.substring(0, fnFolder.indexOf("package.json"))
```
Comment (bcomnes) [inline src/install-deps.js]: Is this redundant with the glob?

File: bin.js [hunk 3]
```diff
@@ -50,9 +51,16 @@ if (argv.help || !sourceArg || !destArg) {
...
 const dest = path.resolve(process.cwd(), destArg)
 
-zipIt.zipFunctions(source, dest, { skipGo: !argv['zip-go'] })
-.then(console.log)
-.catch(err => {
+/* Recursively loop through functions + function folders and install deps */
+installFunctionDeps(source).then(() => {
+  /* Then zip em up */
+  zipIt.zipFunctions(source, dest, { skipGo: !argv['zip-go'] })
+  .then(console.log)
```
Comment (bcomnes) [inline bin.js]: Duplicate promise chain path? Maybe just return the promise from the arrow function and handle the chain in a non-nested manner?

Comment (bcomnes) [reply bin.js]: Unless there is a reason to do it this way and I'm not seeing it.

File: src/install-deps.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+const path = require("path");
+const glob = require("glob");
+const cp = require("child_process");
+const parallelLimit = require("run-parallel-limit");
```
Comment (bcomnes) [inline src/install-deps.js]: If you prefer to use a promise version of this too, thats fine with me.  Just throwing that out there.

File: src/install-deps.js (hunk 1 above)
Comment (DavidWells) [reply src/install-deps.js]: yeah it ignores folders without `package.json`

File: src/install-deps.js (hunk 2 above)
Comment (DavidWells) [reply src/install-deps.js]: IDK you wrote this 😁

Comment (DavidWells) [reply src/install-deps.js]: It works. I know that much

File: bin.js (hunk 3 above)
Comment (DavidWells) [reply bin.js]: Flattened it!

File: src/install-deps.js (hunk 4 above)
Comment (DavidWells) [reply src/install-deps.js]: Not familiar with `p-` libs. If you want to update go for it.

I promisified the main API

File: src/install-deps.js (hunk 2 above)
Comment (bcomnes) [reply src/install-deps.js]: OH! I see, its just the dirname.  Yorp.

**Comment by bcomnes** [comment]
To capture the issue here: running this in buildbot isn't currently working because its run outside of build context.

Here are possible solutions:

- Refactor the build environment to set up and maintain a build environment across the different build steps in https://github.com/netlify/buildbot/blob/master/bot/stage.go#L13-L24 . This is the best solution because it gives clear separate between steps and would help reduce complexity in https://github.com/netlify/build-image
- Move the function prep step into the build step inside of run_build_functions.  This would be fairly simple, but we would need to inject a tmp folder for ZISI to work in.
- Something else... Ideas?

While I strongly advocate for the first option, the second option is a decent compromise without adding additional complexity (more of a reshuffle of where it is now.)

**Comment by sw-yx** [comment]
lets just npm all the way for now? stop the immediate pain

**Comment by bcomnes** [comment]
It shouldn't be difficult to accommodate yarn if we move ZISI into the build environment directly. That is the change that is necessary.

**Comment by sw-yx** [comment]
you do have the most context here. really up to you.

**Comment by bcomnes** [comment]
@biilmann would love your input on this as well. 

- The alternative would be a bash version of the nested npm/yarn install logic inside of https://github.com/netlify/build-image
- Dependency caching is limited/complex and unimplemented with this path.

**Comment by sw-yx** [comment]
sweet! shall we ship it?

**Comment by bcomnes** [comment]
I want to get some more sign-off on it today, but yeah, I think its the best approach we've come up with so far.  Also, it's it a non-breaking change from the module API perspective. You can opt in to the install flow by swapping the method you consume.

**Comment by sw-yx** [comment]
fixed a merge conflict

**Comment by bcomnes** [comment]
WE should release this a major or minor: (0.4.0 or 1.0.0).  The bin shouldn't be consumed anywhere else as a library, so minor is safe, or maybe we just call this 1.0.

You can play with this right now on `xenial` by setting `ZISI_VERSION` to `0.4.0-2`.  

Screen Shot 2019-05-08 at 4 19 26 PM

File: src/install-deps.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    return fnFolder.substring(0, fnFolder.indexOf("package.json"))
+  })
+}
+
+function verifyInstaller(type) {
+  if (type !== 'npm' && type !== 'yarn') {
+    throw new Error('Must check npm or yarn')
+  }
+  return new Promise((resolve, reject) => {
+    cp.exec(`which ${type}`, (error, stdout, stderr) => {
```
Comment (bcomnes) [inline src/install-deps.js]: Need to double check if this works on windows.

Comment (ehmicky) [reply src/install-deps.js]: This should not work on Windows. There's [`node-which`](https://github.com/npm/node-which) but this is a bad hack that is not maintained.

**Comment by ehmicky** [comment]
This PR is currently very outdated compared to the latest `master`.

Netlify Build [automatically installs dependencies](https://github.com/netlify/build/blob/master/packages/build/src/install/functions.js) in Netlify Functions before calling `zip-it-and-ship-it`. Our users will get this feature as they opt-it in the Netlify Build beta.

Therefore I think we probably should close this PR? 

@erquhart

**Comment by erquhart** [comment]
Go for it 👍

**Pull request closed without merging**